packages/components/src/drop-zone/index.js

```diff
@@ -2,67 +2,190 @@
+import { useState, useRef } from '@wordpress/element';
 import { upload, Icon } from '@wordpress/icons';
+import { getFilesFromDataTransfer } from '@wordpress/dom';
+import { useRefEffect } from '@wordpress/compose';
 
-/**
- * Internal dependencies
- */
-import { Context, INITIAL_DROP_ZONE_STATE } from './provider';
+function getDragEventType( { dataTransfer } ) {
+	if ( dataTransfer ) {
+		// Use lodash `includes` here as in the Edge browser `types` is implemented
+		// as a DomStringList, whereas in other browsers it's an array. `includes`
+		// happily works with both types.
+		if (
+			includes( dataTransfer.types, 'Files' ) ||
+			getFilesFromDataTransfer( dataTransfer ).length > 0
+		) {
+			return 'file';
+		}
+
+		if ( includes( dataTransfer.types, 'text/html' ) ) {
+			return 'html';
+		}
+	}
+
+	return 'default';
+}
+
+function useFreshRef( value ) {
+	const ref = useRef();
+	ref.current = value;
+	return ref;
+}
 
 export function useDropZone( {
-	element,
 	onFilesDrop,
 	onHTMLDrop,
-	onDrop,
+	onDrop: _onDrop,
+	onDragStart,
+	onDragEnter: _onDragEnter,
+	onDragLeave: _onDragLeave,
+	onDragEnd,
 	isDisabled,
-	withPosition,
-	__unstableIsRelative: isRelative = false,
+	onDragOver: _onDragOver,
 } ) {
-	const dropZones = useContext( Context );
-	const [ state, setState ] = useState( INITIAL_DROP_ZONE_STATE );
-
-	useEffect( () => {
-		if ( ! isDisabled ) {
-			const dropZone = {
-				element,
-				onDrop,
-				onFilesDrop,
-				onHTMLDrop,
-				setState,
-				withPosition,
-				isRelative,
-			};
-			dropZones.add( dropZone );
-			return () => {
-				dropZones.delete( dropZone );
-			};
-		}
-	}, [
-		isDisabled,
-		onDrop,
-		onFilesDrop,
-		onHTMLDrop,
-		withPosition,
-		isRelative,
-	] );
+	const onFilesDropRef = useFreshRef( onFilesDrop );
+	const onHTMLDropRef = useFreshRef( onHTMLDrop );
+	const onDropRef = useFreshRef( _onDrop );
+	const onDragStartRef = useFreshRef( onDragStart );
+	const onDragEnterRef = useFreshRef( _onDragEnter );
+	const onDragLeaveRef = useFreshRef( _onDragLeave );
+	const onDragEndRef = useFreshRef( onDragEnd );
+	const onDragOverRef = useFreshRef( _onDragOver );
 
-	const { x, y, ...remainingState } = state;
-	let position = null;
+	return useRefEffect(
+		( element ) => {
+			if ( isDisabled ) {
+				return;
+			}
 
-	if ( x !== null && y !== null ) {
-		position = { x, y };
-	}
+			let isDraggingGlobally = false;
+
+			element.setAttribute( 'data-wp-is-drop-target', 'true' );
+
+			const { ownerDocument } = element;
+
+			function onDragEnter( event ) {
+				event.preventDefault();
+				if ( onDragEnterRef.current ) {
+					onDragEnterRef.current( event );
+				}
+			}
+
+			const throttledOnDragOver = throttle( ( event ) => {
+				if ( onDragOverRef.current ) {
+					onDragOverRef.current( event );
+				}
+			}, 200 );
+
+			function onDragOver( event ) {
+				// Only call onDragOver for the innermost hovered drag zones.
```

Comment:
I was wondering how this would be replicated from the previous way it worked 😄

Reply:
Yeah, this is so much simpler than checking element rectangles. :)